### PR TITLE
Enhance README to take advantage of `terraform.tf`

### PR DESCRIPTION
**Pull request by jmcgeheeiv**
## what
* Add `terraform.tf` to the usage instructions

## why
* The current usage does not account for `terraform.tf`, which is actually quite brilliant functionality

## references
* Personal experience using this over the last few months

### CODE REVIEW

**Comment by jmcgeheeiv** [comment]
I added a destroy procedure. Using this module this way works very smoothly. Very nice architecture!

**Comment by schollii** [comment]
I was just thinking the other day of doing something like this, I didn't know such functionality was already in place just not documented!

However I'm wondering since it is so far undocumented can we make the following changes before this goes "public":

- rename those 3 input variables, they are unnecessarily verbose and clash with rest. Given the context of this module, it would be sufficient to call them just backend_file_path etc.
- the path to the generated file should probably default to root module path, ie almost always the module that ultimately needs a backend. So the count 0/1 test should be whether the filename has been defined, not whether the path has been defined.
- I think you have to empty bucket after state moved locally, before terraform destroy, otherwise you get an error. 

I would be happy to submit some PR for same changes in other repos if that helps, I don't remember seeing any of them advertise this approach either.

**Comment by osterman** [comment]
This is an interesting approach!

**Comment by schollii** [comment]
pretty sure you will have to set force_destroy to true before doing the final apply since the bucket is versioned

**Comment by jmcgeheeiv** [comment]
Yes, I encountered that last time I did `terraform destroy`. I'm also reconsidering whether the backend file should be added to `.gitignore`, or whether it should be committed to Git. I want to try the workflow a few more times, but right now I'm away from this part of the project now. Still, the basic workflow is quite good. It is CloudPosse that deserves credit for that, @osterman.

**Comment by osterman** [comment]
Sorry we haven’t been able to review this yet. We’ve been slammed with new projects. 

Has anyone else had a chance to test this out?

**Comment by jmcgeheeiv** [comment]
No rush, @osterman. In fact, I will change this to WIP because I still want to check into the issues I mentioned in my last comment.

**Comment by schollii** [comment]
@jmcgeheeiv even though the file is auto-generated I think it should be committed because this way, if there is a change in the tf state module, or you make a change in input values that you didn't realize would affect it, you will be able to see the change with git diff, whereas without having committed it, you will only see what the plan tells you has changed, there will be no source to check for diff to help understand cause of change.

Is there any disadvantage? If a timestamp or other attribute in the generated file changed at every apply, that would be a good reason to gitignore it, but that's not the case.

**Comment by jmcgeheeiv** [comment]
@schollii, in my most recent pass through the workflow I have made these changes:
1. Specify file name `./backend.tf`. `terraform.tf` is not descriptive.
1. Commit `./backend.tf` to git
1. Do something more forceful to remove the state bucket. @schollii, could you help by describing the exact command you recommend?

Shall I implement these changes in the documentation?

**Comment by schollii** [comment]
I gather this will work on v0.16.0? Pretty sure I had issues with 0.17.0 (#47). It seems like it's only the docs that are changing. I believe the workflow is that after you have deleted the backend.tf and done terraform init so state is all local, you then have to set the force_destroy to true in the bucket in backend-bootstrap.tf (the file that instantiates the tfstate-backend module and creates bucket/dynamodb table), then terraform destroy. However I cannot verify that right now, but if you don't do that, then during terraform destroy you will get an error that the bucket has contents. After I deleted manually from AWS console a couple weeks ago, the destroy worked.

**Comment by schollii** [comment]
@jmcgeheeiv I have verified the workflow and documented at https://github.com/cloudposse/terraform-aws-tfstate-backend/issues/44#issuecomment-640954116 since PR's are harder to find. HTH.

**Comment by Gowiem** [comment]
@jmcgeheeiv Definitely like the idea of getting this merged. I personally didn't know about this functionality and it sounds much preferred! Weighing in on a couple things: 

1. `backend.tf` > `terraform.tf` 👍 
1. Committing `backend.tf` to git -- I only see negatives in not doing this. I think this is the right call. 

Happy to get this PR approved, merged, and cut for you as soon as you and @schollii suss out the final changes you want to make to the README. Thanks gents!

**Comment by jmcgeheeiv** [comment]
Modified without question @schollii's procedure in #44. Because he seems like such a trustworthy chap.

@schollii, @Gowiem, please scrutinize.

**Comment by jmcgeheeiv** [comment]
Today I had an opportunity to destroy a deployment, and I found an error in my destroy procedure (the value of `force_destroy`).

I also added @schollii as a contributor. I hope you do not mind.

**Comment by Gowiem** [comment]
@jmcgeheeiv Sorry to do it to ya, but it looks like there are some merge conflicts. Mind fixing and we'll get this merged?

File: README.md [hunk 1]
```diff
@@ -101,39 +101,94 @@ We literally have [*hundreds of terraform modules*][terraform_modules] that are
...
     module "another_module" {
       source = "....."
     }
    ```
+   Module inputs `terraform_backend_config_file_path` and
+   `terraform_backend_config_file_name` control the name of the backend
+   definition file. Note that when `terraform_backend_config_file_path` is
+   empty (the default), no file is created.
+
+1. `terraform init`. This downloads Terraform modules and providers.
```
Comment (aknysh) [inline README.md:138]: why all steps are numbered `1`? :) 
I see it was like that before, but maybe now is a good time to fix that.

File: README.md [hunk 2]
```diff
@@ -101,39 +101,94 @@ We literally have [*hundreds of terraform modules*][terraform_modules] that are
...
+1. In `main.tf`, change the `terraform_state_backend` module arguments as
+   follows:
+   ```hcl
+    module "terraform_state_backend" {
+        ...
+      terraform_backend_config_file_path = ""
+      force_destroy                      = true
+    }
+    ```
+1. `terraform apply -target module.terraform_state_backend -auto-approve`.
```
Comment (aknysh) [inline README.md:182]: all steps are numbered as `1`

**Comment by schollii** [comment]
@aknysh markdown doesn't require sequence of numbers just first one must be '1'. See https://www.markdownguide.org/basic-syntax/#ordered-lists. 

However the readme is understandable without rendering if we use sequence of numbers, for small lists it's manageable.

**Comment by RothAndrew** [comment]
IMO 1s is preferable, since you dont need to reorder them if you add an item in the middle of the list

**Comment by aknysh** [comment]
> IMO 1s are preferable, since you dont need to reorder them if you add an item in the middle of the list

ah yes, the user will see it as an unordered list, it's just in the code itself we have `1`.

**Comment by RothAndrew** [comment]
They will see it as an ordered list

**Comment by aknysh** [comment]
> They will see it as an ordered list

**Comment by Gowiem** [comment]
@aknysh Did you mean to close this? I would think we would still want to get this merged. Mind if I reopen?

**Comment by aknysh** [comment]
> @aknysh Did you mean to close this? I would think we would still want to get this merged. Mind if I reopen?

oh sorry, clicked the wrong button, reopening

**Comment by jmcgeheeiv** [comment]
The [GitHub style guide specifies that `1.` be used](https://github.com/style-guides/Markdown#list)

**Comment by Gowiem** [comment]
@jmcgeheeiv @schollii Thanks for the contributions and wading through the updates! This is just a docs update, but still... Released as [0.18.1](https://github.com/cloudposse/terraform-aws-tfstate-backend/releases/tag/0.18.1).

**Pull request merged**